torch/utils/_sympy/expr_ranges.py

```diff
@@ -0,0 +1,178 @@
+import enum
+import sys
+from typing import Optional, Tuple
+
+import torch
+from torch import fx
+from .numbers import int_oo
+
+
+class SymRel(enum.Enum):
+    EQ = 0
+    NE = 1
+    LT = 3
+    LE = 4
+    GT = 5
+    GE = 6
+
+
+class SymExprRange:
+    '''
+    Represents the range of a symbolic expression.
+    Stores exact value (if exists), min/max bounds, whether the bounds are in/exclusive,
+    inequalities, and whether the expression is size-like.
+
+    Supports refinement calls eq(), ne(), lt(), gt(), and set_is_size().
+    Handles interactions, for example:
+        - gt(3), gt(5), gt(4) -> gt(5)
+        - lt(3, strict=False), ne(3) -> lt(3, strict=True)
+        - lt(3, strict=False), gt(3, strict=False) -> eq(3)
+
+    Assumes the calls are coherent (i.e. some value exists that satisfies all calls),
+    for example gt(5), lt(4) will silently fail.
+
+    Refinement calls return None if the call had no effect, and a tuple of (SymRel, val)
+    if the range is modified. The returned SymRel may differ from the called method,
+    for example lt(3, strict=False), gt(3, strict=False) returns (SymRel.EQ, 3)
+    since the range is made static.
```

Comment:
I'm a bit unsure why this concept is needed at the moment.

Reply:
Isn't this `ValueRanges.__and__`?